pkg/bucketaccessrequest/bucketaccessrequest.go

```diff
@@ -57,13 +58,24 @@ func (b *bucketAccessRequestListener) Add(ctx context.Context, obj *v1alpha1.Buc
 	return nil
 }
 
+// Update is called in response to a change to BucketAccessRequesst. At this point
+// BucketAccess cannot be changed once created as the Provisioner might have already acted upon the create BucketAccess and created the backend Bucket Credentials
+// Changes to Protocol, Provisioner, BucketInstanceName, BucketRequest cannot be allowed. Best is to delete and recreate a new BucketAccessRequest
+// Changes to ServiceAccount, PolicyActionsConfigMapData and Parameters should be considered in lieu with sidecar implementation
 func (b *bucketAccessRequestListener) Update(ctx context.Context, old, new *v1alpha1.BucketAccessRequest) error {
 	glog.V(1).Infof("Update called for BucketAccessRequest %v", old.Name)
+	if (old.ObjectMeta.DeletionTimestamp == nil) &&
+		(new.ObjectMeta.DeletionTimestamp != nil) {
+		// BucketAccessRequest is being deleted, check and remove finalizer once BA is deleted
+		return b.removeBucketAccess(ctx, new)
+	}
 	return nil
 }
 
-func (b *bucketAccessRequestListener) Delete(ctx context.Context, obj *v1alpha1.BucketAccessRequest) error {
-	glog.V(1).Infof("Delete called for BucketAccessRequest %v", obj.Name)
+// Delete is in response to user deleting a BucketAccessRequest. The call here will respond by deleting a BucketAccess Object.
+func (b *bucketAccessRequestListener) Delete(ctx context.Context, bucketAccessRequest *v1alpha1.BucketAccessRequest) error {
+	glog.V(1).Infof("Delete called for BucketAccessRequest %v", bucketAccessRequest.Name)
+
```

Comment:
In [L92](https://github.com/kubernetes-sigs/container-object-storage-interface-controller/pull/29/files#diff-967e432aed6a8d88ba1f533d2aa548e67f3cc310d734ccdb66deeeab2aaef788R92) below do we want to check that the existing BA belongs to this BAR?

Reply:
In [L97](https://github.com/kubernetes-sigs/container-object-storage-interface-controller/pull/29/files#diff-967e432aed6a8d88ba1f533d2aa548e67f3cc310d734ccdb66deeeab2aaef788R97) below should err "IsMissing" be checked. IOW, a non-nil error could be an intermittent API failure and COSI should retry. How it is code COSI won't retry.

Reply:
In [L103](https://github.com/kubernetes-sigs/container-object-storage-interface-controller/pull/29/files#diff-967e432aed6a8d88ba1f533d2aa548e67f3cc310d734ccdb66deeeab2aaef788R103) below if the BR is nil and/or the err is "IsMissing" why do we return an error? Isn't this likely the user deleted their BAR before COSI processed it?  If we do want to return an error in this case I think a better value would be something like `("Expected BAR %s/%s is missing and assumed to be deleted.", BAR.namespace, BAR.name)`

Reply:
In [L103](https://github.com/kubernetes-sigs/container-object-storage-interface-controller/pull/29/files#diff-967e432aed6a8d88ba1f533d2aa548e67f3cc310d734ccdb66deeeab2aaef788R103) below I think we need more robust code here. If we cannot get the BR we don't necessarily have an "invalidBucketRequest". It could be an intermittent API failure and a retry will fix that. It could be that the BR was deleted and we have a problem since the BAR requires a BR. In either case I think we need a more accurate error msg.

Reply:
In [L112](https://github.com/kubernetes-sigs/container-object-storage-interface-controller/blob/e62441a746154332bb2818d58a0f8ad5959a6641/pkg/bucketaccessrequest/bucketaccessrequest.go#L112) below is there a max on how long we wait? Is this built into some max value the controller is created with?

Reply:
In L92 below do we want to check that the existing BA belongs to this BAR?  - Yes
In L97 below should err "IsMissing" be checked. IOW, a non-nil error could be an intermittent API failure and COSI should retry. How it is coded COSI won't retry.  - Any error would trigger retry and this error will cause retry
In L103 below - BAR is not present we simply retry until we backoff? 
L112 - retry is configurable.